### PR TITLE
lightning: add ExternalEngineConfig to create an external engine

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/45719

Problem Summary:

### What is changed and how it works?

Add ExternalEngineConfig. local backend CloseEngine will use ExternalEngineConfig to create an external engine, and then local backend can import it with the UUID.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/local/local_test.go [hunk 1]
```diff
@@ -2145,61 +2145,85 @@ func TestCtxCancelIsIgnored(t *testing.T) {
...
+
+	done := make(chan struct{})
+	go func() {
+		for i := 0; i < 5; i++ {
+			jobs = append(jobs, <-jobToWorkerCh)
+			testJobWg.Done()
+		}
+	}()
+	go func() {
+		err2 := local.CloseEngine(
```
Comment (lance6716) [inline br/pkg/lightning/backend/local/local_test.go:2211]: this is the expected usage

File: br/pkg/lightning/backend/local/local.go [hunk 2]
```diff
@@ -1113,7 +1160,6 @@ func (local *Backend) generateAndSendJob(
 	jobWg *sync.WaitGroup,
 ) error {
 	logger := log.FromContext(ctx)
-	// TODO(lance6716): external engine should also support it
 	localEngine, ok := engine.(*Engine)
```
Comment (tangenta) [inline br/pkg/lightning/backend/local/local.go:1163]: Can we remove this type conversion?

**Review by tangenta — COMMENTED** [review]
Rest LGTM

File: br/pkg/lightning/backend/local/local.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/lightning/backend/local/local.go:1163]: This is too get smaller region to improve concurrency when use dynamic region. I think external engine does not support to split again. If we want the region size smaller than the cluster configuration, we can use a smaller value when create RangeSplitter?

Comment (lance6716) [reply br/pkg/lightning/backend/local/local.go:1163]: Oh the problem is when use dynamic region, we should use large range size to split & scatter, but smaller range size to import. So we must split region again, I think we can postpone the fix to future PR.

Comment (tangenta) [reply br/pkg/lightning/backend/local/local.go:1163]: Maybe move this assignment to `if` statement to make it local to prevent it from being misused.

**Review by tangenta — APPROVED** [review]

File: br/pkg/lightning/backend/external/engine.go [hunk 3]
```diff
@@ -177,13 +185,13 @@ func (e *Engine) createMergeIter(ctx context.Context, start kv.Key) (*MergeKVIte
 	return iter, nil
 }
 
-// KVStatistics returns the total kv size and total kv length.
+// KVStatistics returns the total kv size and total kv count.
 func (e *Engine) KVStatistics() (totalKVSize int64, totalKVLength int64) {
```
Comment (wjhuang2016) [inline br/pkg/lightning/backend/external/engine.go:189]:
```suggestion
func (e *Engine) KVStatistics() (totalKVSize int64, totalKVCount int64) {
```

File: br/pkg/lightning/backend/local/local.go [hunk 4]
```diff
@@ -735,12 +740,21 @@ func (local *Backend) lockAllEnginesUnless(newState, ignoreStateMask importMutex
 
 // Close the local backend.
 func (local *Backend) Close() {
-	allEngines := local.lockAllEnginesUnless(importMutexStateClose, 0)
+	allEngines := make([]common.Engine, 0, 8)
+	for _, e := range local.externalEngine {
+		allEngines = append(allEngines, e)
```
Comment (wjhuang2016) [inline br/pkg/lightning/backend/local/local.go:745]: Why not close it directly?

File: br/pkg/storage/local.go [hunk 5]
```diff
@@ -139,7 +139,13 @@ func (l *LocalStorage) Open(_ context.Context, path string) (ExternalFileReader,
 
 // Create implements ExternalStorage interface.
 func (l *LocalStorage) Create(_ context.Context, name string, _ *WriterOption) (ExternalFileWriter, error) {
-	file, err := os.Create(filepath.Join(l.base, name))
+	filename := filepath.Join(l.base, name)
+	dir := filepath.Dir(filename)
+	err := os.MkdirAll(dir, 0750)
```
Comment (wjhuang2016) [inline br/pkg/storage/local.go:144]: Why do you change it?

File: br/pkg/lightning/backend/local/local.go (hunk 4 above)
Comment (lance6716) [reply br/pkg/lightning/backend/local/local.go]: https://github.com/pingcap/tidb/pull/46636/commits/4a8925702248bbb4c5af0c99574b117a13c0b266

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**